File: codes/frontend/src/lib/ticket.test.ts
import { describe, it, expect } from 'vitest'
import { base64urlnopad } from '@scure/base'
import { generateOrgKeypair } from './orgKeypair'
import {
  createTicketPayload,
  signTicket,
  decodeTicket,
  verifyTicket,
  TICKET_TTL_SECONDS,
} from './ticket'

const POLL = '0x1111111111111111111111111111111111111111' as const
const NOW = 1_000_000

describe('ticket', () => {
  it('round-trips {p, n, e} through sign → decode', () => {
    const kp = generateOrgKeypair()
    const payload = createTicketPayload(POLL, NOW)
    const enc = signTicket(payload, kp.privKey)
    const decoded = decodeTicket(enc)
    expect(decoded.p).toBe(POLL)
    expect(decoded.n).toBe(payload.n)
    expect(decoded.e).toBe(NOW + TICKET_TTL_SECONDS)
  })

  it('verifies a valid, unexpired ticket', () => {
    const kp = generateOrgKeypair()
    const enc = signTicket(createTicketPayload(POLL, NOW), kp.privKey)
    const res = verifyTicket(enc, kp.pubKey, NOW)
    expect(res.valid).toBe(true)
    if (res.valid) expect(res.ticket.p).toBe(POLL)
  })

  it('rejects an expired ticket with reason "expired"', () => {
    const kp = generateOrgKeypair()
    const enc = signTicket(createTicketPayload(POLL, NOW), kp.privKey)
    const res = verifyTicket(enc, kp.pubKey, NOW + TICKET_TTL_SECONDS + 1)
    expect(res.valid).toBe(false)
    if (!res.valid) expect(res.reason).toBe('expired')
  })

  it('rejects a ticket signed by a different key with reason "badSig"', () => {
    const signer = generateOrgKeypair()
    const attacker = generateOrgKeypair()
    const enc = signTicket(createTicketPayload(POLL, NOW), signer.privKey)
    const res = verifyTicket(enc, attacker.pubKey, NOW)
    expect(res.valid).toBe(false)
    if (!res.valid) expect(res.reason).toBe('badSig')
  })

  it('rejects a tampered preimage with reason "badSig"', () => {
    const kp = generateOrgKeypair()
    const enc = signTicket(createTicketPayload(POLL, NOW), kp.privKey)
    const wire = base64urlnopad.decode(enc)
    wire[0] = wire[0]! ^ 0xff // flip a pollId byte; signature no longer matches
    const tampered = base64urlnopad.encode(wire)
    const res = verifyTicket(tampered, kp.pubKey, NOW)
    expect(res.valid).toBe(false)
    if (!res.valid) expect(res.reason).toBe('badSig')
  })

  it('rejects malformed input without throwing', () => {
    const kp = generateOrgKeypair()
    expect(verifyTicket('AAAA', kp.pubKey, NOW)).toEqual({ valid: false, reason: 'malformed' })
    expect(verifyTicket('not valid base64 @@@', kp.pubKey, NOW)).toEqual({
      valid: false,
      reason: 'malformed',
    })
  })

  it('rejects an invalid pollId at payload creation', () => {
    expect(() => createTicketPayload('0xabc', NOW)).toThrow()
  })
})

File: codes/app/packages/core_crypto/test/crypto/ticket_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:convert/convert.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:core_crypto/crypto/ticket.dart';

/// Cross-client byte-compatibility for the signed ticket. Golden preimage/wire
/// come from codes/frontend/src/lib/ticket.ts (the relayer verifies the same
/// layout). A pass proves a Dart-signed ticket verifies on the relayer and a
/// relayer/web-signed ticket verifies here.
void main() {
  final fixture = jsonDecode(
    File('test/fixtures/cross_client_vectors.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  final ed = fixture['ed25519'] as Map<String, dynamic>;
  final tv = fixture['ticket'] as Map<String, dynamic>;
  final seedHex = ed['seedHex'] as String;
  final pubKey = ed['pubKey'] as String;
  final fields = tv['fields'] as Map<String, dynamic>;
  final goldenPreimageHex = tv['preimageHex'] as String;
  final goldenWire = tv['wire'] as String;

  Ticket goldenTicket() => Ticket(
        p: fields['p'] as String,
        n: fields['n'] as String,
        e: fields['e'] as int,
      );

  group('ticket — golden cross-client vectors', () {
    test('ticketPreimage matches the 32-byte TS preimage', () {
      expect(hex.encode(ticketPreimage(goldenTicket())), goldenPreimageHex);
    });

    test('signTicket reproduces the exact TS wire string', () {
      expect(signTicket(goldenTicket(), seedHex), goldenWire);
    });

    test('decodeTicket round-trips the golden wire', () {
      final t = decodeTicket(goldenWire);
      expect(t.p, fields['p']);
      expect(t.n, fields['n']);
      expect(t.e, fields['e']);
    });
  });

  group('ticket — verifyTicket', () {
    test('valid + unexpired (now < e)', () {
      final r = verifyTicket(goldenWire, pubKey, now: (fields['e'] as int) - 1);
      expect(r.valid, isTrue);
      expect(r.ticket!.p, fields['p']);
    });

    test('expired when now >= e', () {
      final r = verifyTicket(goldenWire, pubKey, now: fields['e'] as int);
      expect(r.valid, isFalse);
      expect(r.reason, TicketInvalidReason.expired);
    });

    test('badSig for a different public key', () {
      const otherPub =
          '0000000000000000000000000000000000000000000000000000000000000001';
      final r = verifyTicket(goldenWire, otherPub, now: (fields['e'] as int) - 1);
      expect(r.valid, isFalse);
      expect(r.reason, TicketInvalidReason.badSig);
    });

    test('badSig for a tampered preimage byte', () {
      // Flip the first pollId byte; signature no longer matches.
      final bytes = base64Url.decode(base64Url.normalize(goldenWire));
      bytes[0] = bytes[0] ^ 0xff;
      final tampered = base64Url.encode(bytes).replaceAll('=', '');
      final r = verifyTicket(tampered, pubKey, now: (fields['e'] as int) - 1);
      expect(r.valid, isFalse);
      expect(r.reason, TicketInvalidReason.badSig);
    });

    test('malformed input does not throw', () {
      expect(verifyTicket('AAAA', pubKey, now: 1).reason,
          TicketInvalidReason.malformed);
      expect(verifyTicket('not valid base64 @@@', pubKey, now: 1).reason,
          TicketInvalidReason.malformed);
    });
  });

  group('ticket — createTicketPayload', () {
    test('sets expiry = now + ttl and normalises pollId', () {
      final t = createTicketPayload(
        '0x1111111111111111111111111111111111111111',
        1000,
      );
      expect(t.e, 1000 + ticketTtlSeconds);
      expect(t.p, '0x1111111111111111111111111111111111111111');
      expect(t.n.length, 16); // 8 bytes hex
    });

    test('rejects an invalid pollId', () {
      expect(() => createTicketPayload('0xabc', 1000), throwsArgumentError);
    });
  });
}
